library(shiny)
library(shinydashboard)
library(ape)
library(tidyverse)
library(plotly)

Tree <- get(load("clade.tree.Rsave")) %>% ladderize
DD <- get(load('ttomes_clades.Rsave')) %>% mutate(GenSp=paste(Genus, Species, `Culture ID`, sep=" "))
II <- get(load("metric_info.Rsave"))
PP <- get(load("pipeline.Rsave"))

Tip.count <- DD %>% group_by(Clade) %>% count
TipC <- setNames(Tip.count$n, Tip.count$Clade)
Tip.labs <- DD %>% group_by(Clade) %>% count %>% ungroup %>% select(n) %>% unlist %>% unname
Tree$tip.label <- paste(Tree$tip.label, " (", TipC[Tree$tip.label], ")", sep="")
DD1 <- get(load('ttomes_clades.Rsave')) %>% 
  dplyr::select(`Num reads F and R`:`Num nuclear read pairs`, `Percent merged`:`BUSCOs percent`)

header  <- dashboardHeader(titleWidth = 300, title="Diatom Transcriptomes")
sidebar <- dashboardSidebar(
  sidebarMenu(
    menuItem("Phylogeny navigator", tabName = "Phylogeny", icon = icon("th")),
    menuItem("Metric relationships", icon = icon("th"), tabName = "Metrics"),
    menuItem("Species cards", icon = icon("th"), tabName = "Species"),
    menuItem("Pipeline", icon = icon("th"), tabName = "Pipeline"),
    menuItem("About", icon = icon("th"), tabName = "About")
  )
)
body    <- dashboardBody(
  tabItems(
    tabItem(tabName = "Phylogeny",
            fluidRow(
              box(title = "Navigate by clade", solidHeader = TRUE, collapsible = TRUE, width = 3, status = "primary",
                  helpText("Select a region by brushing on the phylogeny. 
                            The number of transcriptomes per lineage is in parentheses after the lineage name."),
                  plotOutput("plot1", height = "900px", brush = "plot_brush")
                  ),
              uiOutput("box1", width=7),
              infoBoxOutput("ibox1", width=3),
              uiOutput("box2", width=7),
              infoBoxOutput("ibox2", width = 3)
            )
          ),
    tabItem(tabName = "Metrics",
            fluidRow(
              box(width = 3, status = 'primary', solidHeader = TRUE, title = "Select metrics to compare",
                checkboxGroupInput("check_metrics", label = "Select two metrics", choices = colnames(DD1), selected = colnames(DD1)[1:2])
                ),
            uiOutput("box3", width=7),
            infoBoxOutput("ibox3", width = 3),
            infoBoxOutput("ibox4", width = 3)
              )
    ),
    tabItem(tabName = "Species",
            h3("Widgets tab content")
    ),
    tabItem(tabName = "Pipeline",
            box(width = 12, status = 'primary', solidHeader = TRUE, title = "Transcriptome processing pipeline",
                tableOutput("pipe")
            )
    ),
    tabItem(tabName = "About",
            h3("Project"),
            h3("Evaluating the contributions of horizontally transferred bacterial genes and 
               endogenous duplication events to the diversification of diatoms"),
            tags$div(HTML("<font size='2'>The goal of this project is to identify the sources of novel diatom traits, 
                          which have contributed to their extraordinary metabolic and species-level diversity. 
                          Some traits are encoded by genes acquired from distantly related bacteria, an entirely different 
                          kingdom in the tree of life. Other genes appear to have been generated by processes acting within diatom genomes. 
                          This project will generate genome-scale data for 250 phylogenetically disparate diatom species. 
                          These data will be used to: (1) infer evolutionary relationships, (2) identify the sources - 
                          whether intrinsic or extrinsic - of new genes, and (3) correlate the pattern and 
                          timing of gene acquisitions with the origins of novel traits.</font>"),
                     tags$br(),
                     tags$br(),
                     HTML("<font size='2'>This study would result in the training of a post-doctoral researcher, undergraduates and graduate students. 
                          With an estimated 200,000 species, diatoms are one of the most diverse groups of marine and freshwater phytoplankton. 
                          These microscopic 'plants' play a major role in the global cycling of carbon and oxygen, producing the oxygen for 
                          one of every five human breaths. They manufacture the fatty acids that make their way up the food chain, eventually 
                          into fish, which humans harvest for fish oils. Diatoms possess a broad range of novel traits that, together, have 
                          facilitated their rise to prominence in the world's oceans. Ultimately, this project will reveal how nature has assembled
                          the mosaic genomes of one of the world's most diverse and ecologically important lineages.</font>"),
                     tags$br(),
                     tags$br(),
                     tags$a(href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1353152", HTML("<font size='3'>(NSF Award DEB-1353152)</font>"))),
            tags$hr(),
            h3("People"),
            tags$div(
              HTML("<font size='3'>Andy Alverson</font>"), 
              tags$a(href="https://alversonlab.com", icon("home", "fa-2x")), 
              tags$a(href="https://twitter.com/sinkingdiatom", icon("twitter", "fa-2x")),
              tags$a(href="https://github.com/andrewalverson", icon("github", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Genome/transcriptome assembly, phylogenomics, whole genome duplication, horizontal gene transfer, differential gene expression (PI)</font>"),
              tags$br(), tags$br(),
              HTML("<font size='3'>Norm Wickett</font>"), 
              tags$a(href="http://faculty.wcas.northwestern.edu/wickett/", icon("home", "fa-2x")), 
              tags$a(href="https://twitter.com/NormWickett", icon("twitter", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Genome/transcriptome assembly, phylogenomics, whole genome duplication, horizontal gene transfer (PI)</font>"),
              tags$br(), tags$br(),
              HTML("<font size='3'>Elizabeth Ruck</font>"),
              tags$a(href="https://alversonlab.com", icon("home", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Diatom culturing and wet-lab guru, organellar genes/genomes, whole genome duplication (Post-doc)</font>"),
              tags$br(), tags$br(),
              HTML("<font size='3'>Matt Parks</font>"), 
              tags$a(href="https://www.researchgate.net/profile/Matthew_Parks2", icon("home", "fa-2x")),
              tags$a(href="https://github.com/mparkscbg", icon("github", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Genome/transcriptome assembly, phylogenomics, whole genome duplication, horizontal gene transfer (Post-doc)</font>"),
              tags$br(), tags$br(),
              HTML("<font size='3'>Teofil Nakov</font>"),
              tags$a(href="http://teofil.discindo.org", icon("home", "fa-2x")), 
              tags$a(href="https://twitter.com/teofilnakov", icon("twitter", "fa-2x")),
              tags$a(href="https://github.com/teofiln", icon("github", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Phylogenomics, whole genome duplication, differential gene expression, this Shiny App (Post-doc)</font>"),
              tags$br(), tags$br(),
              HTML("<font size='3'>Kala Downey</font>"),
              tags$a(href="https://alversonlab.com", icon("home", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Wet-lab, culturing, differential gene expression (PhD student)</font>"),
              tags$br(), tags$br(),
              HTML("<font size='3'>Colton Kessenich</font>"),
              tags$a(href="https://alversonlab.com", icon("home", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Transcriptome assembly & annotation (PhD student)</font>"),
              tags$br(), tags$br(),
              HTML("<font size='3'>Rachel Ungar</font>"),
              tags$a(href="https://alversonlab.com", icon("home", "fa-2x")),
              tags$br(),
              HTML("<font size='2'>Diatom-bacteria associations (Undergraduate student)</font>")
            ),
            tags$hr(),
            h3("Papers"),
            HTML("Kessenich, C. R., Ruck, E. C., Schurko, A. M., Wickett, N. J., & Alverson, A. J. (2014). 
                 Transcriptomic insights into the life history of bolidophytes, the sister lineage to diatoms. 
                 Journal of phycology, 50(6), 977-983."),
            tags$br(), tags$br(),
            HTML("Parks, M. B., Wickett, N. J., & Alverson, A. J. (2017). Signal, Uncertainty, and Conflict 
in Phylogenomic Data for a Diverse Lineage of Microbial Eukaryotes (Diatoms, Bacillariophyta). 
                 Molecular biology and evolution, 35(1), 80-93."),
            tags$br(), tags$br(),
            HTML("Parks, M., Nakov, T., Ruck, E. C., Wickett, N. J., & Alverson, A. J. (2017). 
                   Phylogenomics reveals an extensive history of genome duplication in diatoms (Bacillariophyta). 
                   bioRxiv, 181115. Forthcoming in American Journal of Botany"),
            tags$hr(),
            h3("Data and code"),
            tags$hr()
    )
  )
)

ui <- dashboardPage(
  header,
  sidebar,
  body
)

server <- function(input, output) {

#####
  # TAB 1
  
  #------------------------------------------------#
  # stuff for the phylogeny
  
  # functon to plot the tree  
  plot.tree <- function(Tree, labelOffset=.2){
    par(lend=2, mar=rep(.5,4))
    plot.phylo(Tree, font=1, edge.width=4, edge.color="grey60", cex=1, label.offset=labelOffset, use.edge.length = F)
  }
  
  # render the plot
  output$plot1 <- renderPlot({
    plot.tree(Tree, labelOffset=1)
    par(lwd=4)
  })
  
  # get coordinates of species on the plot
  getTreeInfo <- reactive({
    plot.tree(Tree)
    plotinfo <- get("last_plot.phylo", envir = .PlotPhyloEnv)
    tips_xy <- data.frame(Tip=Tree$tip.label, XX=plotinfo$xx[1:Ntip(Tree)], YY=plotinfo$yy[1:Ntip(Tree)])
    return(tips_xy)
  })
  
  # capture the brushed area
  whichArea <- reactive({
    QQ <- getTreeInfo()
    wanted_tip <- QQ$Tip[QQ$YY >= input$plot_brush$ymin & QQ$YY <= input$plot_brush$ymax]
    wanted_tip <- map(wanted_tip, function(x) stringr::str_split(x, pattern = " ")[[1]][1]) %>% unname
    return(wanted_tip)
  })

  #------------------------------------------------#
  # first metric
  
  # reactive plot box
  output$box1 <- renderUI({
    if (length(whichArea()) < 1) {
      Title <- "Please select a region on the phylogeny."
    } else {
      X <- paste(whichArea(), collapse=", ")  
      Title <- paste("Data for species in the lineage(s) of ", X, " diatoms." , sep="") 
    }
    
    box(title = Title, 
        solidHeader = TRUE, collapsible = TRUE, width = 6, status = 'primary',
        selectInput("metric1", "Select first metric", selected = colnames(DD1)[1], choices = colnames(DD1)),
        actionButton("refresh1", "Refresh", icon = icon("refresh")),
        helpText("Click 'Refresh' to update the plot."),
        # validate(
        #   need(!is.null(input$plot_brush), 'Please select a region of the phylogeny!')#,
        #   # need(input$refresh1, 'Please refresh to update the plot!')
        #   ),
        plotlyOutput("m1")
    )
  })
  
  # reactive info box
  output$ibox1 <- renderInfoBox({
    infoBox(
      title = input$metric1,
      subtitle = paste(II[names(II)==input$metric1]),
      width = 2,
      color = "light-blue",
      icon = icon("info-circle")
    )
  })
  
  # capture clicks, subset data, make plot
  pl1 <- eventReactive({
      input$plot_click
      input$refresh1 
    }, {
    
    DD2 <- DD %>% 
      filter(Clade %in% whichArea()) %>%
      dplyr::select(Clade, GenSp, yVar=matches(input$metric1))
    
    pp <- ggplot(DD2, aes(x=GenSp, y=yVar)) +
      geom_point(pch=21, size=3, fill="orange", stroke=.3) +
      coord_flip() +
      ylab(as.character(input$metric1)) +
      xlab("") +
      theme_minimal()
    
    pp <- ggplotly(pp)
    pp$x$data[[1]]$text <- paste("Species", ": ", DD2$GenSp, "<br>",
                                 "Clade", ": ", DD2$Clade, "<br>",
                                 input$metric1, ": ", format(DD2$yVar, big.mark=","), sep="")
    
    return(pp)
  })
  
  # render the plot
  output$m1 <- renderPlotly({
    pl1()
  })
  
#------------------------------------------------#

pl2 <- eventReactive({
  input$plot_click
  input$refresh3 
  #  input$refreshAll
}, {
  
  DD2 <- DD %>% 
    filter(Clade %in% whichArea()) %>%
    dplyr::select(Clade, GenSp, yVar=matches(input$metric2))
  
  pp <- ggplot(DD2, aes(x=GenSp, y=yVar)) +
    geom_point(pch=21, size=3, fill="orange", stroke=.3) +
    coord_flip() +
    ylab(as.character(input$metric2)) +
    xlab("") +
    theme_minimal()
  
  pp <- ggplotly(pp)
  pp$x$data[[1]]$text <- paste("Species", ": ", DD2$GenSp, "<br>",
                               "Clade", ": ", DD2$Clade, "<br>",
                               input$metric1, ": ", format(DD2$yVar, big.mark=","), sep="")
  
  return(pp)
})

output$m2 <- renderPlotly({
  pl2()
})

# reactive plot box
output$box2 <- renderUI({
  if (length(whichArea()) < 1) {
    Title <- "Please select a region on the phylogeny."
  } else {
    X <- paste(whichArea(), collapse=", ")  
    Title <- paste("Data for species in the lineage(s) of ", X, " diatoms." , sep="") 
  }
  
  box(title = Title, 
      solidHeader = TRUE, collapsible = TRUE, width = 6, status = 'primary',
      selectInput("metric2", "Select second metric", selected = colnames(DD1)[1], choices = colnames(DD1)),
      actionButton("refresh3", "Refresh", icon = icon("refresh")),
      helpText("Click 'Refresh' to update the plot."),
      plotlyOutput("m2")
  )
})

output$ibox2 <- renderInfoBox({
  infoBox(
    title = input$metric2,
    subtitle = paste(II[names(II)==input$metric2]),
    width = 2,
    color = "light-blue",
    icon = icon("info-circle")
  )
})

# END TAB 1
### --- ###

#####
# TAB 2

# reactive scatter plot
output$box3 <- renderUI({
  if (length(input$check_metrics) != 2) {
    Title <- "Please select two metrics."
  } else {
    X <- paste(input$check_metrics, collapse=" and ")  
    Title <- paste("Relationship between ", X, "." , sep="") 
  }
  
  box(title = Title, 
      solidHeader = TRUE, collapsible = TRUE, width = 6, status = 'primary',
      plotlyOutput("m3")
  )
})

pl3 <- reactive({
  
  DD2 <- DD %>% 
    dplyr::select(Clade, GenSp, xVar=matches(input$check_metrics[1]), yVar=matches(input$check_metrics[2]))
  
  pp <- ggplot(DD2, aes(x=xVar, y=yVar)) +
    geom_point(pch=21, size=2, fill="orange", stroke=.3) +
    geom_smooth(method=lm)+
    ylab(as.character(input$check_metrics[2])) +
    xlab(as.character(input$check_metrics[1])) +
    theme_minimal()
  
  pp <- ggplotly(pp)
  pp$x$data[[1]]$text <- paste("Species", ": ", DD2$GenSp, "<br>",
                               "Clade", ": ", DD2$Clade, "<br>",
                               input$check_metrics[1], ": ", format(DD2$xVar, big.mark=","), "<br>",
                               input$check_metrics[2], ": ", format(DD2$yVar, big.mark=","),
                               sep="")
  
  return(pp)
})

# render the plot
output$m3 <- renderPlotly({
  pl3()
})

output$ibox3 <- renderInfoBox({
  infoBox(
    title = input$check_metrics[1],
    subtitle = paste(II[names(II)==input$check_metrics[1]]),
    width = 2,
    color = "light-blue",
    icon = icon("info-circle")
  )
})

output$ibox4 <- renderInfoBox({
  infoBox(
    title = input$check_metrics[2],
    subtitle = paste(II[names(II)==input$check_metrics[2]]),
    width = 2,
    color = "light-blue",
    icon = icon("info-circle")
  )
})


# END TAB 2
### --- ###

#####
# TAB 3


# output$box4 <- renderUI({
#   Title <- paste(input$species, collapse="")  
#   box(title = Title, 
#       solidHeader = TRUE, collapsible = FALSE, width = 12, status = 'primary',
#       # picture
#       # picture
#       # classification
#       # raw data
#       # assembly
#       # quality
#       # external links
#         # algebase
#         # diatombase
#         # CCMP
#         # NCBI
#         # MMETSP
#         # ProtistCentral
#   )
# })
# output$tab1 <- renderTable({
#   classification
# })

# END TAB 3
### --- ###

#####
# TAB 4

output$pipe <- renderTable(PP, striped = TRUE, bordered = TRUE, spacing = "s")

# END TAB 4
### --- ###

#####
# TAB 5

# END TAB 5
### --- ###
}

shinyApp(ui, server)
